### ISSUE TITLE
Skip all reduce globally unused parameters in DDP

### ISSUE THREAD

**Issue by zhaojuanmao** [pytorch/pytorch#42183]
We can provide an option to users to skip all reduce globally unused parameters in DDP. For this option, we assume locally unused parameters are the same across the ranks. It can save communication time when these unused parameters size is large

**Comment by mcarilli** [comment]
Because of parameter bucketing, this incurs some performance tradeoffs (it isn't an obvious win), and I'm curious if the !/$ is worthwhile.

We'd have to either change the bucket structure before backward if prepare_for_backward discovers a new set of unused parameters in a particular iteration, or keep the same bucket structure and copy any partially-filled buckets to fully-packed sub-buckets then allreduce the sub-buckets.  The former incurs additional CPU work and the latter incurs additional GPU work within each process.  In other words, if only a few parameters are unused, inserting logic to not allreduce unused parameters may not deliver a speedup over blindly allreducing zeros for those parameters.  Are there known use cases where not allreducing unused parameters provides substantial benefits?

**Comment by zhaojuanmao** [comment]
@mcarilli you are right about the partial bucketing! There are some cases requested by users where unused parameters could be large, and skip communicating them may potentially improve the performance.

Regarding partial bucketing, one thing we can further simplify is: we skip communicating the bucket if parameters in the bucket are all unused, this is mostly helpful when unused parameters sizes are large. and it is simple enough. 

how do you think?

**Comment by mcarilli** [comment]
Do you imagine this option will be used only with `find_unused_parameters`?  If not, it may become tricky:  If a bucket isn't receiving gradients, during backward (without find_unused_parameters) Reducer has no way to tell if the bucket expects to receive gradients or not.  It will wait for that bucket to receive gradients, other buckets (which do receive gradients) will be serialized behind it, and none of them will be allreduced.

**Comment by zhaojuanmao** [comment]
@mcarilli yes, this option will be used only with find_unused_parameters, otherwise it will not work properly as you mentioned. So we probably need to design the API carefully. Otherwise, we have to wait until "Modularize DistributedDataParallel" is implemented, which allow DDP package is configurable.  see https://github.com/pytorch/pytorch/issues/37002

**Comment by rohan-varma** [comment]
From a user perspective, I assume to implement the above suggestion (if find_unused_params=True and assume that locally unused param implies globally unused) we would need to add a flag to the DDP API which users would set to True if this assumption worked for them? Also, I was wondering if it might make sense (from a perf perspective) to make all globally unused parameters share the same bucket, which might simplify some of this logic?
